packages/cli/lib/charm.ts

```diff
@@ -217,35 +218,78 @@ export async function linkCharms(
 ): Promise<void> {
   const manager = await loadManager(config);
 
-  const sourceCharm = await manager.get(sourceCharmId, false);
-  if (!sourceCharm) {
-    throw new Error(`Source charm "${sourceCharmId}" not found`);
+  // Try to get source as a charm first, then as an arbitrary cell ID
+  let sourceCell: Cell<any>;
+  let sourceIsCharm = false;
+  try {
+    const sourceCharm = await manager.get(sourceCharmId, true); // Run the charm to ensure it has results
+    sourceCell = sourceCharm;
+    sourceIsCharm = true;
+    
+  } catch (error) {
+    // If manager.get() fails (e.g., "recipeId is required"), try as arbitrary cell ID
+    try {
+      sourceCell = await manager.getCellById({ "/": sourceCharmId });
+      sourceIsCharm = false;
+    } catch (cellError) {
+      throw new Error(`Source "${sourceCharmId}" not found as charm or cell`);
+    }
   }
 
-  const targetCharm = await manager.get(targetCharmId, false);
-  if (!targetCharm) {
-    throw new Error(`Target charm "${targetCharmId}" not found`);
+  // Try to get target as a charm first, then as an arbitrary cell ID
+  let targetCell: Cell<any>;
+  let targetIsCharm = false;
+  try {
+    const targetCharm = await manager.get(targetCharmId, true); // Run the charm to ensure it exists
+    targetCell = targetCharm;
+    targetIsCharm = true;
+  } catch (error) {
+    // If manager.get() fails (e.g., "recipeId is required"), try as arbitrary cell ID
+    try {
+      targetCell = await manager.getCellById({ "/": targetCharmId });
+      // Check if this cell is actually a charm by looking at the charms list
+      const charms = manager.getCharms().get();
+      const isActuallyCharm = charms.some(charm => {
+        const id = charmId(charm);
```

Comment:
should this be `getCharmIdSafe` from #1285? I guess we'd just skip it if it lacks a charm id